File: 01 - rocket-equation/JS/main.js
const fs = require("fs");

(async () => {
    const masses = await readFile('input.txt');
    const total = process(masses);
    console.log(total);
})();

function fuelCounterUpper(moduleMass) {
    let fuelMass = Math.floor(moduleMass / 3);
    fuelMass = fuelMass - 2;
    if (fuelMass <= 6) return fuelMass;
    return fuelMass + fuelCounterUpper(fuelMass);
}

function validateInput(input) {
    if (isNaN(input)) {
        console.log('encountered not a number in file');
        exit();
    } else {
        return input;
    }
}

async function readFile(path) {
    return new Promise((resolve, reject) => {
        return fs.readFile(path, 'utf8', (err, data) => {
            if (err) reject(err);
            else {
                resolve(data.trim().split('\n'));
            }
        });
    });
}

function process(masses) {
    return masses.map(mass => validateInput(mass))
        .map(mass => fuelCounterUpper(mass))
        .reduce((a, b) => { return a + b }, 0);
}

File: 01 - rocket-equation/Java/Main.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        var masses = readFile("input.txt");
        var total = process(masses);
        total.ifPresentOrElse(System.out::println, () -> System.out.println("error"));
    }

    static Optional<Integer> process(final List<Integer> masses) {
        return masses.stream().map(Main::calculateFuelCost).reduce(Integer::sum);
    }

    static int calculateFuelCost(final int moduleMass) {
         var fuelMass = (int)(Math.floor(moduleMass/3) - 2);
         if (fuelMass <= 6) return fuelMass;
         return fuelMass + calculateFuelCost(fuelMass);
    }

    static List<Integer> readFile(final String path) {
        File file = new File(path);
        var list = new ArrayList<Integer>();
        try {
            Scanner scan = new Scanner(file);
            while (scan.hasNextInt()) {
                list.add(scan.nextInt());
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        return list;
    }
}
